src/adapter/src/optimize.rs

```diff
@@ -132,6 +133,12 @@ where
             }
             Ok(result) => result,
             Err(panic) => {
+                // A panic during optimization is always a bug; log an error so we learn about it.
+                tracing::error!(
+                    "caught a panic during query optimization: {panic}\n{}",
+                    Backtrace::force_capture()
+                );
+
                 let msg = format!("unexpected panic during query optimization: {panic}");
```

Comment:
I didn't attempt to change the wording of this error message in this PR because that's user facing and warrants some SQL council input.